File: go/tests/power/ade7953_test/main.go
//go:build linux && !tinygo

// ADE7953 hardware-in-loop test for the Linux host.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/power"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, _ := strconv.Atoi(envOr("I2C_BUS", "1"))
	addr64, _ := strconv.ParseUint(envOr("I2C_ADDR", "0x38"), 0, 8)

	tr, err := connection.NewI2CConnection(bus, uint8(addr64), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer tr.Close()

	chip, err := power.NewADE7953Full(tr, 251.0, 30.0)
	if err != nil {
		fmt.Fprintln(os.Stderr, "init:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, ok bool) {
		if ok {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	if v, err := chip.Voltage(); err == nil {
		check("voltage non-negative", v >= 0.0)
	}
	if i, err := chip.Current(); err == nil {
		check("current non-negative", i >= 0.0)
	}
	if p, err := chip.ActivePower(); err == nil {
		check("activePower finite", p > -1.0e6)
	}
	if e, err := chip.ActiveEnergy(); err == nil {
		check("activeEnergy finite", e > -1000.0)
	}
	if lp, err := chip.LinePeriod(); err == nil {
		check("linePeriod positive", lp > 0.0)
	}

	if err := chip.Reset(); err == nil {
		if v, err := chip.Voltage(); err == nil {
			check("voltage after reset", v >= 0.0)
		}
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

File: rust/tests/power/ade7953_test/src/main.rs
use linux_embedded_hal::{Delay, I2cdev};
use periph::chips::power::Ade7953Minimal;

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x38);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut delay = Delay;
    let mut ade = Ade7953Minimal::new(dev, addr, 251.0, 30.0, &mut delay).expect("init ADE7953");

    let mut passed = 0i32;
    let mut failed = 0i32;
    check_true!(ade.voltage().unwrap_or(-1.0) >= 0.0, "voltage non-negative", passed, failed);
    check_true!(ade.current().unwrap_or(-1.0) >= 0.0, "current non-negative", passed, failed);
    check_true!(ade.active_power().unwrap_or(-1.0e6) > -1.0e6, "activePower finite", passed, failed);
    check_true!(ade.active_energy().unwrap_or(-1000.0) > -1000.0, "activeEnergy finite", passed, failed);
    check_true!(ade.line_period().unwrap_or(0.0) > 0.0, "linePeriod positive", passed, failed);

    ade.reset(&mut delay).expect("reset");
    check_true!(ade.voltage().unwrap_or(-1.0) >= 0.0, "voltage after reset", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}